doc/notes/188_enhance_logging_int.adoc

```diff
@@ -0,0 +1,215 @@
+= Enhance Logging
+
+xtUML Project Implementation Note
+
+== 1 Abstract
+
+This note describes the implementation of the `log4cplus` logging.
+
+== 2 Introduction and Background
+
+See <<dr-4>> for requirements, analysis and design of the enhanced
+logging.
+
+As part of fulfilling these requirements, the MASL C{plus}{plus} model
+compiler <<dr-2>> was extended.  An overview will be provided here.
+
+The big win with this logging is that Reporting requirements are fulfilled
+together and consistently with Logging requirements.  Reporting is a
+subset of Logging and uses the same infrastructure and conventions.  By
+having a hierarchical set of loggers, the application can be instrumented
+to configure data acquisition for Reporting and for Logging in a flexible
+manner which minimises performance impacts.  This provides some resilience
+to Reporting requirements changes allowing the PV to be re-configured to
+supply more (or less) information to the Reporting back end.
+
+It is noted that as of March 2024, this work is incomplete.  Structured
+logging, deprecation of the Reporting terminator and other details remain.
+However, the work is ready for a first merge into main to facilitate
+integration issues with test and benchmarking.
+
+=== 2.1 Examples
+
+Through configuration changes alone, the following can/could be
+accomplished:
+
+* We can log to a Kafka topic or the syslog daemon instead of a file (or
+  to both).
+* Within a single logger (e.g. 'pv.aeo') We can send INFO level logs to one
+  file (or topic) and send DEBUG level logs to a different file (or topic).
+  This is done with the logger filtering capability of `log4cplus`.
+  In our current usage, this makes it simple to divide Debug and
+  Information/Warning traffic into different destinations.
+* By extending our logger hierarchy, we can enable future debug of
+  individual classes, operations, functions and/or states.  For example, since
+  we are instrumenting the Sequence Verification domain HappyJob class
+  JobInProgress state machine state as 'pv.sv.happyjob.jobinprogress', we
+  enable the ability to debug only JobInProgress if we desire.  This
+  additional fine-grained logger hierarchy brings no additional configuration
+  burden, since 'pv.sv.happyjob.jobinprogress' can inherit from 'pv.sv' in
+  production mode.  It is only when we desire to debug that state do we
+  configure a specific logger at the lower level.  No code change is
+  required.
+* Many features for formatting the log messages with file names, operation
+  names, timestamps, line numbers, etc. are now available.
+* These hierarchical loggers (e.g. 'system.domain.class.state') are
+  complemented with horizontal loggers (e.g. 'system.benchmarking').  (We
+  will use 'pv.benchmarking.jm', 'pv.benchmarking.aeo' and
+  'pv.benchmarking.sv' to empower the fine-grained concept horizontally as
```

Comment:
Is it pv.benchmarking or pv.benchmark? There seems to be some inconsistency between this note and the 188_enhance_logging_ant.adoc note.

Reply:
In the logging config file, I used benchmark.  I will change it in this note.